rasa/core/actions/forms.py

```diff
@@ -656,3 +691,21 @@ async def deactivate(self, *args: Any, **kwargs: Any) -> List[Event]:
         """Deactivates form."""
         logger.debug(f"Deactivating the form '{self.name()}'")
         return []
+
+    async def _activate_loop(
+        self,
+        output_channel: "OutputChannel",
+        nlg: "NaturalLanguageGenerator",
+        tracker: "DialogueStateTracker",
+        domain: "Domain",
+        events: List[Event],
+    ) -> List[Event]:
+        temp_tracker = tracker.copy()
+
+        if not await self.is_activated(output_channel, nlg, tracker, domain):
+            default_activation_events = self._default_activation_events()
+            temp_tracker.update_with_events(default_activation_events, domain)
+            events += default_activation_events
+            events += await self.activate(output_channel, nlg, temp_tracker, domain)
+
+        return events
```

Comment:
I am not sure what this function does, especially the operation on the temp tracker. Can you explain?

Reply:
If you look into the `LoopAction.run()` which gets inherited by `FormAction`, this method does three things:
1. activates the form
2. allows the form to collect info (or `do`)
3. deactivates the form once conditions for deactivation are met

It is only after all these steps are done, does the MessageProcessor update the tracker with events (which includes the vital `ActiveLoop` event which sets the `tracker.active_loop_name` to the form name). This is what gets checked by the default action `action_extract_slots` [here](https://github.com/RasaHQ/rasa/blob/e8c0f8495a3960a0feac64fbf3f2fb1c75cc49ce/rasa/core/actions/action.py#L1028) when coming across slots with mapping conditions.

Because this update of the tracker by the MessageProcessor happens **too late**, I had to override a part of the superclass `run` method, namely the `activation` part which I extracted and used a temp tracker instead that gets updated with `ActiveLoop` event in L707 - this is a default activation event. This way ACTION_EXTRACT_SLOTS can fill a slot with a mapping condition during activation, because the `temp_tracker.active_loop_name` will be set to the current form name and will pass the check in the default action `_verify_mapping_conditions()`.

Let me know if any of this is not clear enough 🙏🏼

Reply:
OK I think that I understand, thanks a lot for detailing this. Something that was confusing to was the use of a temp tracker. Basically it is mainly used here as a way to compute events that need to actually be set in the real tracker.